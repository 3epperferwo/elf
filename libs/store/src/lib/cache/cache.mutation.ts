import { CacheState } from './cache.state';
import { Reducer } from '@eleanor/store';

export function setCache<S extends CacheState>(cache: CacheState['$cache']): Reducer<S> {
  return state => {
    return {
      ...state,
      $cache: cache
    };
  };
}
