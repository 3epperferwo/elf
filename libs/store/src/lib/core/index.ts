export { Store, Reducer, StoreValue } from './store';
export { stateFactory } from './state-factory';
export { select, head, distinctUntilArrayItemChanged } from './operators';
export { getStore, getRegistry } from './registry';
export { withProps } from '../props/props.state';
