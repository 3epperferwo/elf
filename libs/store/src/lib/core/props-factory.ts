import {select} from './operators';
import {OperatorFunction} from 'rxjs';
import {EmptyConfig, State} from './state';
import {Reducer} from './store';
import {capitalize, isFunction, isObject} from './utils';

export function propsFactory<T extends Record<any, any>,
  K extends keyof T = T extends Record<infer Key, any> ? Key : never>(key: K, initialValue: T[K]) {
  const normalizedKey = capitalize(key as string);

  return {
    [`with${normalizedKey}`](value = initialValue) {
      return {
        state: {
          [key]: value,
        },
        config: undefined,
      };
    },
    [`set${normalizedKey}`](value: any) {
      return function (state: any) {
        const newVal = isFunction(value) ? value(state) : value;

        return {
          ...state,
          [key]: newVal,
        };
      };
    },
    [`update${normalizedKey}`](value: any) {
      return function (state: any) {
        const newVal = isFunction(value) ? value(state) : value;

        return {
          ...state,
          [key]: isObject(value) ? {
            ...state[key],
            ...newVal
          } : newVal
        };
      };
    },
    [`reset${normalizedKey}`]() {
      return function (state: any) {
        return {
          ...state,
          [key]: initialValue,
        };
      };
    },
    [`select${normalizedKey}`]() {
      return select((state: any) => state[key]);
    },
    [`get${normalizedKey}`](state: any) {
      return state[key];
    },
  } as unknown as {
    [P in | `get${Capitalize<string & K>}`
      | `set${Capitalize<string & K>}`
      | `update${Capitalize<string & K>}`
      | `reset${Capitalize<string & K>}`
      | `select${Capitalize<string & K>}`
      | `with${Capitalize<string & K>}`]:

    P extends `set${Capitalize<string & K>}`
      ? <S extends T>(value: S[K] | ((state: S) => S[K])) => Reducer<S> :
      P extends `update${Capitalize<string & K>}`
        ? <S extends T>(value: Partial<S[K]>) => Reducer<S>
        : P extends `get${Capitalize<string & K>}`
        ? <S extends T>() => S[K]
        : P extends `select${Capitalize<string & K>}`
          ? <S extends T>() => OperatorFunction<S, S[K]>
          : P extends `reset${Capitalize<string & K>}`
            ? <S extends T>() => Reducer<S>
            : (initialValue?: T[K]) => State<T, EmptyConfig>;
  };
}
