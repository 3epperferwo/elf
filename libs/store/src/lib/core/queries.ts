import { OperatorFunction, pipe } from 'rxjs';
import { distinctUntilChanged, map } from 'rxjs/operators';

export function select<T, R>(mapFn: (state: T) => R): OperatorFunction<T, R> {
  return pipe(
    map(mapFn),
    distinctUntilChanged()
  );
}

export function head<
  T extends any[],
  Item = T extends (infer I)[] ? I : never
  >(): OperatorFunction<T, Item> {
  return map((arr: T) => arr[0] as Item);
}
