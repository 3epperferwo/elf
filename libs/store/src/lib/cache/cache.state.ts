import { EmptyConfig, State } from '../core/state';

export type CacheState = {
  $cache: 'none' | 'partial' | 'full';
}

export function withCache(): State<CacheState, EmptyConfig> {
  return {
    state: {
      $cache: 'none'
    },
    config: {}
  };
}
