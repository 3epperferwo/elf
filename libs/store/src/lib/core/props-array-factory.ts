import { arrayAdd, arrayRemove, arrayToggle } from '../array';
import { capitalize } from './utils';
import { propsFactory } from './props-factory';
import { Reducer } from './store';

export function propsArrayFactory<
  T extends any[],
  K extends string,
  PropState extends { [Key in K]: T }
>(key: K, options: { initialValue: T; config?: any; idKey?: keyof T[0] }) {
  const normalizedKey = capitalize(key as string);
  const base = propsFactory<T, K, PropState>(key, options);

  return {
    ...base,
    [`add${normalizedKey}`](value: any) {
      return function (state: any) {
        return {
          ...state,
          [key]: arrayAdd(state[key], value),
        };
      };
    },
    [`remove${normalizedKey}`](id: any) {
      return function (state: any) {
        return {
          ...state,
          [key]: arrayRemove(state[key], id),
        };
      };
    },
    [`toggle${normalizedKey}`](id: any) {
      return function (state: any) {
        return {
          ...state,
          [key]: arrayToggle(state[key], id, { idKey: options?.idKey ?? 'id' }),
        };
      };
    },
  } as unknown as typeof base &
    {
      [P in
        | `remove${Capitalize<K>}`
        | `add${Capitalize<K>}`
        | `toggle${Capitalize<K>}`]: P extends `toggle${Capitalize<K>}`
        ? <S extends PropState>(value: T[0]) => Reducer<S>
        : P extends `add${Capitalize<K>}`
        ? <S extends PropState>(value: T[0]) => Reducer<S>
        : P extends `remove${Capitalize<K>}`
        ? <S extends PropState>(value: T[0]) => Reducer<S>
        : never;
    };
}
