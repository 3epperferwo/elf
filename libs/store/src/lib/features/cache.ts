import {Query, StateOf} from "../core/types";
import {EMPTY, Observable, OperatorFunction} from "rxjs";
import {Reducer, Store, StoreDef} from "../core/store";
import { propsFactory } from "../core/props-factory";
import { select } from "../core/operators";

export type CacheStatus = 'none' | 'partial' | 'full';

export const {
  withCache,
  setCache,
  selectCache,
  resetCache,
  getCache,
  updateCache
} = propsFactory<{ cache: Record<string | number, CacheStatus> }>('cache', {});

export function deleteCacheEntry<S extends StateOf<typeof withCache>>(key: string | number): Reducer<S> {
  return function (state: S) {
    const cache = {...state.cache};

    if (Reflect.deleteProperty(cache, key)) {
      return {
        ...state,
        cache
      }
    }

    return state;
  }
}

export function inCache<S extends StateOf<typeof withCache>>(key: string | number, status?: CacheStatus): Query<S, boolean> {
  return function (state: S) {
    if (status) {
      return state.cache[key] === status;
    }

    return Reflect.has(state.cache, key);
  }
}

export function selectInCache<S extends StateOf<typeof withCache>>(key: string | number, status?: CacheStatus): OperatorFunction<S, boolean> {
  return select(state => inCache(key, status)(state));
}


export function skipWhileCached<T, S extends StateOf<typeof withCache>>(store: Store<StoreDef<S>>, key: string, type?: CacheStatus) {
  return function (source: Observable<T>) {
    if (store.query(inCache(key, type))) {
      return EMPTY;
    }

    return source;
  }
}
