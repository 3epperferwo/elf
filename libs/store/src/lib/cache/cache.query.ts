import { select } from '../core/queries';
import { CacheState } from './cache.state';

export function selectCache<S extends CacheState>() {
  return select<S, S['$cache']>(state => state.$cache);
}
