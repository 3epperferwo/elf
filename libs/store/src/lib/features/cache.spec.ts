import {createState, Store} from '../core';
import {withEntities} from '../entities';
import {deleteCacheEntry, inCache, resetCache, selectCache, selectInCache, updateCache, withCache} from "./cache";

describe('cache', () => {
  it('should work', () => {
    const { state, config } = createState(
      withEntities<{ id: number; title: string }>(),
      withCache()
    );

    const store = new Store({ state, config, name: '' });

    const spy = jest.fn();
    store.pipe(selectCache()).subscribe(spy);
    expect(spy).toHaveBeenCalledWith({});

    store.reduce(updateCache({ todos: 'full' }));

    expect(store.state).toMatchSnapshot('update');

    store.reduce(updateCache({ todo: 'partial' }));

    expect(store.state).toMatchSnapshot('update');

    store.reduce(resetCache());

    expect(store.state).toMatchSnapshot('reset');

    store.reduce(deleteCacheEntry('todo'));

    expect(store.state).toMatchSnapshot('delete todo cache');

    store.reduce(updateCache({ todos: 'full' }));

    expect(store.query(inCache('todos'))).toEqual(true);

    const selectHasCacheSpy = jest.fn();
    store.pipe(selectInCache('todos')).subscribe(selectHasCacheSpy);

    expect(selectHasCacheSpy).toHaveBeenCalledWith(true);

    store.reduce(deleteCacheEntry('todos'));

    expect(selectHasCacheSpy).toHaveBeenCalledWith(false);
  });
});
