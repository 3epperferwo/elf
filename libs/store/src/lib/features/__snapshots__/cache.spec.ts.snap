// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cache should work: delete todo cache 1`] = `
Object {
  "cache": Object {},
  "entities": Object {},
  "ids": Array [],
}
`;

exports[`cache should work: reset 1`] = `
Object {
  "cache": Object {},
  "entities": Object {},
  "ids": Array [],
}
`;

exports[`cache should work: update 1`] = `
Object {
  "cache": Object {
    "todos": "full",
  },
  "entities": Object {},
  "ids": Array [],
}
`;

exports[`cache should work: update 2`] = `
Object {
  "cache": Object {
    "todo": "partial",
    "todos": "full",
  },
  "entities": Object {},
  "ids": Array [],
}
`;
